lib/scripting/objects/vbs-to-wpcemu.spec.ts

```diff
@@ -0,0 +1,136 @@
+/*
+ * VPDB - Virtual Pinball Database
+ * Copyright (C) 2019 freezy <freezy@vpdb.io>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import * as chai from 'chai';
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+import { SinonStub } from 'sinon';
+import { TableBuilder } from '../../../test/table-builder';
+import { Player } from '../../game/player';
+import { Emulator } from '../../emu/wpc-emu';
+import { EmulatorState } from '../../emu/emulator-state';
+
+/* tslint:disable:no-unused-expression no-string-literal */
+chai.use(require('sinon-chai'));
+describe('VpmController integration test', () => {
+
+	const sandbox = sinon.createSandbox();
+	let setSwitchInputSpy: SinonStub<[number, boolean?]>;
+	let setDipSwitchByteSpy: SinonStub<[number]>;
+	let getSolenoidSpy: SinonStub<[number]>;
+	let getLampSpy: SinonStub<[number]>;
+	let getChangedLampsSpy: SinonStub<any>;
+	let getChangedSolenoidsSpy: SinonStub<any>;
+	let getChangedGISpy: SinonStub<any>;
+	let getChangedLEDsSpy: SinonStub<any>;
+
+	beforeEach(() => {
+		setSwitchInputSpy = sandbox.stub(Emulator.prototype, 'setSwitchInput').returns(true);
+		setDipSwitchByteSpy = sandbox.stub(Emulator.prototype, 'setDipSwitchByte');
+		getSolenoidSpy = sandbox.stub(Emulator.prototype, 'getSolenoidState');
+		getLampSpy = sandbox.stub(Emulator.prototype, 'getLampState');
+		getChangedLampsSpy = sandbox.stub(EmulatorState.prototype, 'getChangedLamps');
+		getChangedSolenoidsSpy = sandbox.stub(EmulatorState.prototype, 'getChangedSolenoids');
+		getChangedGISpy = sandbox.stub(EmulatorState.prototype, 'getChangedGI');
+		getChangedLEDsSpy = sandbox.stub(EmulatorState.prototype, 'getChangedLEDs');
```

Comment:
I'm thinking... since you don't use them all in all tests (actually, you only use one per test), wouldn't it be clearer to instantiate them on each test so it's clear when they are used?

Reply:
I addressed that, thanks!

Reply:
I think both variants work - but the suggested one is definitive more explicit!

Reply:
Yeah I think it's definitely easier to have them all at the top while writing the tests.. But more readable having them in the test cases later.

Thanks for moving, will merge when CI is done :)